plugins/inputs/socketstat/socketstat.go

```diff
@@ -0,0 +1,219 @@
+			// a best effort, extend the metrics from the 1st line with the metrics of the 2nd
+			// one, possibly overwriting.
+			for _, word := range words {
+				if ss.ValidValues.MatchString(word) {
+					// kv will have 2 fields because it matched the regexp
+					kv := strings.Split(word, ":")
+					fields[kv[0]], err = strconv.ParseUint(kv[1], 10, 64)
+					if err != nil {
+						ss.Log.Infof("Couldn't parse metric: %s", word)
+						continue
+					}
+				}
+			}
+			if !flushData {
+				ss.Log.Warnf("Found orphaned metrics: %s", words)
+				ss.Log.Warnf("Added them to the last known connection.")
+			}
+			acc.AddFields(ss.Measurement, fields, tags)
+			flushData = false
+		}
+	}
+	if flushData {
+		acc.AddFields(ss.Measurement, fields, tags)
+	}
+	return nil
+}
+
+func getTagsAndState(proto string, words []string, log telegraf.Logger) (map[string]string, map[string]interface{}) {
+	tags := map[string]string{}
+	fields := make(map[string]interface{})
+	tags["proto"] = proto
+	switch proto {
+	case "udp", "raw":
+		words = append([]string{"dummy"}, words...)
+	case "tcp", "dccp", "sctp":
+		fields["state"] = words[0]
+	}
+	switch proto {
+	case "tcp", "udp", "raw", "dccp", "sctp":
+		// Local and remote addresses are fields 3 and 4
+		// Separate addresses and ports with the last ':'
+		localIndex := strings.LastIndex(words[3], ":")
+		remoteIndex := strings.LastIndex(words[4], ":")
+		tags["local_addr"] = words[3][:localIndex]
+		tags["local_port"] = words[3][localIndex+1:]
+		tags["remote_addr"] = words[4][:remoteIndex]
+		tags["remote_port"] = words[4][remoteIndex+1:]
+	case "unix", "packet":
+		fields["netid"] = words[0]
+		tags["local_addr"] = words[4]
+		tags["local_port"] = words[5]
+		tags["remote_addr"] = words[6]
+		tags["remote_port"] = words[7]
+	}
+	var err error
+	fields["recv_q"], err = strconv.ParseUint(words[1], 10, 64)
+        if err != nil {
+                log.Infof("Couldn't read recv_q in: %s", words)
+        }
+	fields["send_q"], err = strconv.ParseUint(words[2], 10, 64)
+	if err != nil {
+		log.Infof("Couldn't read send_q in: %s", words)
+	}
+	return tags, fields
+}
+
+func (ss *Socketstat) Init() error {
+	if len(ss.SocketProto) == 0 {
+		ss.SocketProto = []string{"tcp", "udp"}
+	}
+
+	// Check that ss is installed, get its path
+	ssPath, err := exec.LookPath("ss")
+	if err != nil {
+		return err
+	}
+	ss.cmdName = ssPath
+
+	ss.Measurement = "socketstat"
+
+	if ss.Timeout < config.Duration(time.Second) {
+		ss.Timeout = config.Duration(time.Second)
+	}
+
+	// Initialize regexps to validate input data
+	validFields := "(bytes_acked|bytes_received|segs_out|segs_in|data_segs_in|data_segs_out)"
+	ss.ValidValues = regexp.MustCompile("^" + validFields + ":[0-9]+$")
+	ss.BeginsWithBlank = regexp.MustCompile(`^\s+.*$`)
+
+	return nil
+}
+
+func init() {
+	inputs.Add("socketstat", func() telegraf.Input {
+		return &Socketstat{
+			Lister: socketList,
```

Comment:
Can we please also do this in `Init()`?

Reply:
done.

Reply:
Oh I think my comment was misleading here. Please only move the `Lister = socketList` to `Init()`. We need the call to `inputs.Add` in `init()` (lowercase) as this function is called then on import and registers the plugin to be used... Sorry for the inconvinience!

Reply:
Right, I did that. I wasn't sure there'd be a way to assign a custom value to ss.lister for tests, but found it.